docs/src/references/mqtt-topic-guidelines.md

```diff
@@ -0,0 +1,827 @@
+```
+
+**Pros**
+
+1. Due to the lack of distinction in the topics between parent devices and child devices,
+   it is easier to write code for "the device" irrespective of whether it is deployed as a parent or child.
+1. Minimal payload size as we're only including the `id`s everywhere without repeating any context info.
+
+**Cons**
+
+1. Not easy to differentiate the context(from parent or child) easily from the message.
+1. The `id`s must be unique between all devices and services in a deployment.
+1. Not easy to do subscriptions like: "measurements only from child devices or only from services, excluding parent",
+   without keeping track of all the `id`s of services registered with that child device.
+   Easy wild card subscriptions are not possible at all.
+1. No automatic registration as bootstrapping is mandatory for everything.
+
+## Proposal 3: Namespace extensible topics
+
+This scheme splits the topic levels into two parts:
+* entity/component identification levels
+* data type identification levels
+
+This scheme reserves 4 subtopic levels for entity identification,
+and 2 levels for data type identification such as measurements, alarms, events, commands etc as follows:
+
+```
+te/<name-space-1>/<name-space-2>/<name-space-3>/<name-space-4>/<data-type>/<data-instance-type>
+```
+
+Having 4 levels for entity identification provides enough flexibility
+to better namespace the devices and applications running on it minimizing the risk of `id` conflicts.
+Thin-edge doesn't enforce/dictate how a user uses these namespace levels.
+But there are conventions that can be followed to cover a wide range of use-cases, as follows:
+
+* `te/<device-namespace>/<device-id>/<service-namespace>/<service-id>/...`
+* `te/<device-namespace>/<device-id>/<device-local-namespace>/<device-local-component-id>/...`
+* `te/<entity-namespace>/<entity-id>/<component-namespace>/<component-id>/...`
+* `te/<device-family>/<device-series>/<part-series>/<part-id>/...`
+
+Even for the data type levels, a user is free to define those as they wish.
+But thin-edge has some pre-defined `<data-type>` subtopics for well-known types like:
+`measurements`, `alarms`, `events` and `commands`, on which it enforces some constraints.
+
+**Subtopics for predefined telemetry data**
+
+|Type|Example|
+|------|-------|
+|Measurements|`<entity_id_prefix>/mes/[/<measurement-type>]`|
+|Events|`<entity_id_prefix>/evn/<event-type>`|
+|Alarms|`<entity_id_prefix>/alr/<alarm-type>`|
+
+**Subtopics for predefined commands**
+
+|Type|Example|
+|------|-------|
+|Software List|`<entity_id_prefix>/cmd/software_list`|
+|Software Update|`<entity_id_prefix>/cmd/software_update`|
+|Configuration Snapshot|`<entity_id_prefix>/cmd/config_snapshot`|
+|Configuration Update|`<entity_id_prefix>/cmd/config_update`|
+|Firmware Update|`<entity_id_prefix>/cmd/firmware_update`|
+|Restart[^1]|`<entity_id_prefix>/cmd/restart`|
```

Comment:
The plan was to avoid the req/res notion and switch to a `status` model that is more open. The `id` was already part of the payload. I don't really get the benefit of keeping the `id` in the topic as any component handling a command would most likely subscribe to all commands irrespective of their `id`s. Even if the `id` is in the topic, I see them subscribing to `<op_type>/+` topic to receive all the commands. Obviously we need ordering guarantees for a single operation `id` as well. But we get that even without the `id` in the topic.

Reply:
I'm in favour of having the command (operation) id in the topic name is that it 

* possibility to limit access to specific commands via the [Dynamic Security Plugin](https://mosquitto.org/documentation/dynamic-security/).
* possibility for individual devices to only listen to updates from a specific command and all commands (but yes only subscribing to a specific command instance requires knowing the id upfront)
* commands are more traceable with the command id in the topic (but this is more my opinion than anything else)

Proposed command topic structure:

```text
te/device/main///cmd/update_status/[id]
```

Interested devices/services can subscribe using:

```text
te/device/main///cmd/update_status/+

# Or to a known command id
te/device/main///cmd/update_status/[id]
```

A device/service can register the command capability by publishing to:

```text
te/device/main///cmd/update_status//meta
```

Reply:
This PR only gives the placeholder for the commands (e.g. `te/device/+/service/}/cmd/+` topic). The specifics of how the topic will be used are delegated to this Operation Workflow PR https://github.com/thin-edge/thin-edge.io/pull/2071.

The MQTT api reference document can then be updated afterwards in a different PR.